### PR TITLE
Containerd support

### LINKED ISSUE #779: Add support for containerd
Customer asked to replace Docker by another container runtime for security purposes.

- Docker uses containerd internally and adds another API layer between the kubelet and containerd that is not needed anymore
- Using fluentd / fluent-bit requires to mount /var/lib/docker/containers to the pod, which potentially grants fluetd / fluent-bit access to all running containers. Containerd logs to /var/log/containers and separates logs from other container data.

### PR DESCRIPTION
**Pull request by kron4eg**
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged)*:
Fixes #779 #805 

**Special notes for your reviewer**:

**Optional Release Note**:

```release-note
Containerd support
```

### CODE REVIEW

File: pkg/containerruntime/containerruntime.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+			registryMirrors:    cfg.RegistryMirrors,
+			kubeletVersion:     kubeletVersion,
+		}
+		containerd = &Containerd{
+			insecureRegistries: cfg.InsecureRegistries,
+			registryMirrors:    cfg.RegistryMirrors,
+		}
+	)
+
+	moreThen123, _ := semver.NewConstraint(">= 1.23")
```
Comment (kron4eg) [inline pkg/containerruntime/containerruntime.go:85]: This is the supposed version where dockershim will be removed. But it's unknown for real. If and hen there will be announced another version to remove dockershim, we will change it here.

Comment (xmudrii) [reply pkg/containerruntime/containerruntime.go:85]: As per the dockershim removal KEP, this should be 1.22.

**Review by xmudrii — CHANGES_REQUESTED** [review]

File: pkg/userdata/ubuntu/provider.go [hunk 2]
```diff
@@ -246,19 +188,12 @@ write_files:
...
     sysctl --system
 
-    apt-key add /opt/docker.asc
-    apt-get update
-
 {{- /* Make sure we always disable swap - Otherwise the kubelet won't start'. */}}
-    cp /etc/fstab /etc/fstab.orig
-    cat /etc/fstab.orig | awk '$3 ~ /^swap$/ && $1 !~ /^#/ {$0="# commented out by cloudinit\n#"$0} 1' > /etc/fstab.noswap
-    mv /etc/fstab.noswap /etc/fstab
+    sed -i.orig '/.*swap.*/d' /etc/fstab
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/userdata/ubuntu/provider.go:192]: https://github.com/kubermatic/kubeone/blob/86fab0858897bd9eccc60c6e1a3ad333e27b548b/pkg/scripts/os.go#L30

File: pkg/userdata/coreos/testdata/v1.17.16-vsphere.yaml [hunk 3]
```diff
@@ -141,17 +141,13 @@ systemd:
           --bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf \
           --kubeconfig=/var/lib/kubelet/kubeconfig \
           --config=/etc/kubernetes/kubelet.conf \
-          --network-plugin=cni \
-          --cni-conf-dir=/etc/cni/net.d \
-          --cni-bin-dir=/opt/cni/bin \
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/userdata/coreos/testdata/v1.17.16-vsphere.yaml:146]: yes, those are defaults, + those are docker specific flags, I suppose they will be gone after v1.22+

File: pkg/userdata/helper/helper.go [hunk 4]
```diff
@@ -103,6 +103,95 @@ SystemMaxUse=5G
...
+		criPlugin.Registry.Mirrors["docker.io"] = containerdMirror{
+			Endpoint: registryMirrors,
+		}
+	}
+
+	cfg := containerdConfigManifest{
+		Version: 2,
+		Metrics: &containerdMetrics{
+			// metrics available at http://127.0.0.1:1338/v1/metrics
+			Address: "127.0.0.1:1338",
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/userdata/helper/helper.go:179]: It's default

File: pkg/userdata/helper/helper.go [hunk 5]
```diff
@@ -103,6 +103,95 @@ SystemMaxUse=5G
...
+				},
+			},
+		},
+		Registry: &containerdCRIRegistry{
+			Mirrors: map[string]containerdMirror{
+				"docker.io": {
+					Endpoint: []string{"https://registry-1.docker.io"},
+				},
+			},
+		},
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/userdata/helper/helper.go:160]: I'm not entirely sure if containerd will be merging its defaults with our changes, that's why I've resorted to be explicit about parts that are important.

File: pkg/userdata/centos/provider.go [hunk 6]
```diff
@@ -289,39 +290,16 @@ write_files:
...
 
 - path: /etc/systemd/system/kubelet-healthcheck.service
   permissions: "0644"
   content: |
 {{ kubeletHealthCheckSystemdUnit | indent 4 }}
 
-- path: /etc/systemd/system/docker-healthcheck.service
-  permissions: "0644"
-  content: |
-{{ containerRuntimeHealthCheckSystemdUnit | indent 4 }}
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/userdata/centos/provider.go:305]: I don't think it's worth it. Kubeone works without this nonsense.

File: pkg/userdata/rhel/testdata/kubelet-v1.20-vsphere-mirrors.yaml [hunk 7]
```diff
@@ -369,7 +383,7 @@ write_files:
 - path: /etc/docker/daemon.json
   permissions: "0644"
   content: |
-    {"exec-opts":["native.cgroupdriver=systemd"],"storage-driver":"overlay2","log-driver":"json-file","log-opts":{"max-size":"100m"},"registry-mirrors":["https://registry.docker-cn.com"]}
+    {"exec-opts":["native.cgroupdriver=systemd"],"storage-driver":"overlay2","log-driver":"json-file","log-opts":{"max-size":"100m"}}
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/userdata/rhel/testdata/kubelet-v1.20-vsphere-mirrors.yaml:386]: Apparently no, this is something missing

File: pkg/userdata/helper/testdata/kublet_systemd_unit_version-v1.19.6.golden [hunk 8]
```diff
@@ -21,15 +21,25 @@ ExecStart=/opt/bin/kubelet $KUBELET_EXTRA_ARGS \
...
+||||||| parent of 7a998653 (Update fixtures):pkg/userdata/helper/testdata/kublet_systemd_unit_version-v1.11.3.golden
+  --allow-privileged=true \
+  --network-plugin=cni \
+  --cni-conf-dir=/etc/cni/net.d \
+  --cni-bin-dir=/opt/cni/bin \
+  --cadvisor-port=0 \
+=======
+  --allow-privileged=true \
+  --cadvisor-port=0 \
+>>>>>>> 7a998653 (Update fixtures):pkg/userdata/helper/testdata/kublet_systemd_unit_version-v1.11.3.golden
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/userdata/helper/testdata/kublet_systemd_unit_version-v1.19.6.golden]: it's gone now

File: pkg/userdata/rhel/testdata/kubelet-v1.20-vsphere-mirrors.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/userdata/rhel/testdata/kubelet-v1.20-vsphere-mirrors.yaml]: Fixed

File: pkg/userdata/rhel/testdata/kubelet-v1.20-vsphere-proxy.yaml [hunk 9]
```diff
@@ -369,7 +383,7 @@ write_files:
 - path: /etc/docker/daemon.json
   permissions: "0644"
   content: |
-    {"exec-opts":["native.cgroupdriver=systemd"],"storage-driver":"overlay2","log-driver":"json-file","log-opts":{"max-size":"100m"},"insecure-registries":["192.168.100.100:5000","10.0.0.1:5000"]}
+    {"exec-opts":["native.cgroupdriver=systemd"],"storage-driver":"overlay2","log-driver":"json-file","log-opts":{"max-size":"100m"}}
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/userdata/rhel/testdata/kubelet-v1.20-vsphere-proxy.yaml]: fixed

**Review by xmudrii — APPROVED** [review]
We only need some E2E tests but let's leave them for a follow-up.

/lgtm
/approve
/hold cancel

**Pull request merged**